src/TodosContext.tsx

```diff
@@ -0,0 +1,176 @@
+      break;
+    }
+
+    case (Filter.COMPLETED): {
+      todosCopy = todosCopy.filter(todo => todo.completed);
+      break;
+    }
+
+    default: {
+      break;
+    }
+  }
+
+  return todosCopy;
+}
+
+function filterTodos(todos: Todo[], { filterBy }: FilterParams): Todo[] {
+  return filterBy
+    ? filterTodosByCompleted(todos, filterBy)
+    : todos;
+}
+
+function useLocalStorage<T>(key: string, startValue: T): [T, (v: T) => void] {
+  const [value, setValue] = useState(startValue);
+
+  const save = (newValue: T) => {
+    localStorage.setItem(key, JSON.stringify(newValue));
+    setValue(newValue);
+  };
+
+  return [value, save];
+}
+
+interface ContextProps {
+  todos: Todo[],
+  visibleTodos: Todo[],
+  handleOnAdd: (newQuery: string) => void,
+  handleAllCompletedToggle: (event: React.FormEvent<HTMLInputElement>) => void,
+  handleClearAllCompleted: () => void,
+  handleTodoChange: (newTodo: Todo) => void,
+  handleOnDelete: (todoId: number) => void,
+  isTodosHasCompleted: () => boolean,
+  isEveryTodoCompleted: () => boolean,
+  filterBy: Filter,
+  setFilterBy: (newFilter: Filter) => void,
+}
+
+export const TodosContext = React.createContext<ContextProps>({
+  todos: [],
+  visibleTodos: [],
+  handleOnAdd: () => {},
+  handleAllCompletedToggle: () => {},
+  handleClearAllCompleted: () => {},
+  handleTodoChange: () => {},
+  handleOnDelete: () => {},
+  isTodosHasCompleted: () => false,
+  isEveryTodoCompleted: () => false,
+  filterBy: Filter.ALL,
+  setFilterBy: () => {},
+});
+
+interface ProviderProps {
+  children: React.ReactNode,
+}
+
+export const TodosProvider: React.FC<ProviderProps> = ({ children }) => {
+  const [todos, setTodos] = useLocalStorage<Todo[]>('todos', []);
+  const [filterBy, setFilterBy] = useState(Filter.ALL);
+
+  const visibleTodos = filterTodos(todos, { filterBy });
+
+  const handleOnAdd = (newQuery: string) => {
+    let newTodoId = 1;
+
+    if (todos.length !== 0) {
+      newTodoId = Math.max(...todos.map(todo => todo.id)) + 1;
+    }
+
+    const newTodo: Todo = {
+      id: newTodoId,
+      title: newQuery.trim(),
+      completed: false,
+    };
+
+    const newTodos = [
+      ...todos,
+      newTodo,
+    ];
+
+    setTodos(newTodos);
+  };
+
+  const handleAllCompletedToggle
+    = (event: React.FormEvent<HTMLInputElement>) => {
+      const newTodos = [...todos].map(todo => {
+        return {
+          ...todo,
+          completed: event.currentTarget.checked,
+        };
+      });
+
+      setTodos(newTodos);
+    };
+
+  const handleClearAllCompleted = () => {
+    const activeTodos = [...todos].filter(todo => !todo.completed);
+
+    setTodos(activeTodos);
+  };
+
+  const handleTodoChange = (newTodo: Todo) => {
+    const indexOfOldTodo = todos.findIndex(todo => {
+      return todo.id === newTodo.id;
+    });
+
+    const newTodos = [...todos];
+
+    newTodos.splice(indexOfOldTodo, 1, newTodo);
+
+    setTodos(newTodos);
```

Comment:
wooow) don't do like that. Let's think, how you can edit some item inside the array if you have todo id
```
setTodos(prev => prev.map(item => item.id === newTodo.id ? newTodo : item))
```